Feature/co 745 hours

floppyears opened the pull request.

floppyears committed:
Updates to swagger

CO-745

floppyears committed:
Update swagger to include locations and service

CO-745

jaredkosanovic commented on swagger.yaml:
```diff
@@ -70,7 +70,7 @@ paths:
           type: string
         - in: query
```
What is your collectionFormat for this query parameter? We originally talked about doing "multi" (ex. ?type=dining&type=building) but I wasn't sure if that's what you're implementing.

floppyears replied:
I thought about this during the implementation. I went with just allowing a single type to make the implementation a bit simpler. For now, if needed the consumer can make multiple API calls. Down the road once we migrate to using an ES library, if needed we can support multiple type parameters.

floppyears committed:
Add cultural centers to location search

CO-745

floppyears committed:
Simplify jsonapi objects and use new items from backend

CO-745

floppyears committed:
Refactor to add new fields

CO-745

* Resource class - clean up some of the code
* Mapper - instead of creating objects use json and plain maps

floppyears committed:
Use mapper to get a single location

CO-745

Since ES now contains extra data that we don't want / need in the api
output, the getById method needs to call the mapper class.

floppyears committed:
Refactoring and reuse parent class methods

CO-745

floppyears committed:
Add support for services relationship

CO-745

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/locations/frontend/resources/LocationResource.groovy:
```diff
@@ -73,35 +59,25 @@ class LocationResource extends Resource {
-    @Produces(MediaType.APPLICATION_JSON)
     @Timed
     Response list(@QueryParam('q') String q,
                   @QueryParam('campus') String campus, @QueryParam('type') String type,
```
What if I want to query multiple types? I think we talked about doing something like type=blah&type=building which would be handled as an array of strings. Could you also update the swagger spec to indicate a collectionFormat for the type parameter?

jaredkosanovic commented on src/main/groovy/edu/oregonstate/mist/locations/frontend/resources/LocationResource.groovy:
```diff
@@ -3,33 +3,31 @@ package edu.oregonstate.mist.locations.frontend.resources
 
     public static final ArrayList<String> ALLOWED_CAMPUSES = ["corvallis", "extension"]
-    public static final ArrayList<String> ALLOWED_TYPES = ["building", "dining"]
+    public static final ArrayList<String> ALLOWED_TYPES = ["building", "dining", "cultural-centers"]
```
building and dining is singular, while cultural-centers is plural. Should we just make it "cultural-center" to be consistent?

floppyears replied:
Sure

floppyears committed:
Add services resource class

CO-745

floppyears committed:
Fix extra whitespace reported by codenarc

CO-745

floppyears committed:
Update swagger spec

CO-745

jaredkosanovic commented on swagger.yaml:
```diff
@@ -133,6 +133,115 @@ paths:
+          description: "Successful response"
+          schema:
+            $ref: "#/definitions/ServicesResultObjects"
+        "404":
```
Is this applicable here? If a /services query returned nothing, it will still return a 200 with an empty data array.

floppyears committed:
Update sample configuration file

CO-745

jaredkosanovic commented on swagger.yaml:
```diff
@@ -165,6 +274,13 @@ definitions:
+      data:
+        $ref: "#/definitions/ServicesResourceObject"
+
   ResourceObject:
```
You have ResourceObject/ResultObject and ServicesResourceObject/ServicesResultObject. Could you qualify the locations resource and result object? Ex LocationsResultObject.

jaredkosanovic commented on configuration-example.yaml:
```diff
@@ -47,43 +48,10 @@ locations:
   estype: elastic-search-type
+  esIndexService: elastic-search-services-index
+  estypeService: elastic-search-services-type
+  gatewayUrl: https://api.example.com/v1/
```
I left a comment in the configs repo about this. I was wondering why you're adding back gatewayUrl since we already have endpointUri. I think we changed the code a few months ago to remove gatewayUrl dependencies.

jaredkosanovic commented on swagger.yaml:
```diff
@@ -173,24 +289,42 @@ definitions:
+        items:
+          $ref: "#/definitions/ServicesResourceObject"
+
+  ServicesResourceObject:
```
Should the servicesresourceobject contain a relationships object?

floppyears replied:
Hmm, the jsonapi.org spec wasn't specific as to what was the preferred way to map back to the locations object. I went with a links.location . It seemed cleaner and simple. We could go with the relationships object, I just don't think it makes sense to have both.

jaredkosanovic replied:
You could go with links.related and link that to the locations object. I agree jsonapi isn't too clear though.

jaredkosanovic commented on swagger.yaml:
```diff
@@ -275,18 +448,13 @@ definitions:
         type: string
-        description: Git commit hash of the revision which was built.
-      documentation:
+      location:
```
I don't think we should have this in the links section. I would recommend having the location link in the relationship section similar to how location objects link to services. Also, json api spec only allows links objects to contain "self" and "related" links. You could have the api respond with something like:
```
"relationships": {
    "location": {
        "links":{
            "related": "api.o.e/v1/locations/idofparentlocation"
        }
    }
}

floppyears committed:
Fix to swagger spec based on code review

CO-745

floppyears committed:
Code review fixes

CO-745

floppyears committed:
Update swagger spec

CO-745
